"updated upickle, Http4s, Monocle and scalaZ versions"

ermyas opened the pull request:
updated to the latest versions of upickle, Http4S, Monocle and Scalaz, without disabling scalac options. fixes #9

ermyas committed:
"updated upickle, Http4s, Monocle and scalaZ versions"

ermyas commented:
@beloglazov travis build seems to be failing because http4s 0.9.1 is using java 8 https://github.com/http4s/http4s/issues/158

what do you think about upgrading the client's jvm target to java 8 as well?

ermyas committed:
updated release version number in README

ermyas committed:
updated travis.yml to use jdk8

ermyas committed:
upgraded http4s to 0.9.3, this version might be a fix for issue #8

hieronx commented:
@beloglazov could you merge this PR so I can check whether it fixes #8? Thanks!

bryce-anderson commented:
Quick note: if you're not interested in requiring Java 8, http4s v0.8.6 also contains the fix and targets Java 7. However, all future developments on http4s will require at least Java 8.

ermyas commented:
thanks @bryce-anderson 
@beloglazov I think it might be good to move to Java 8 sooner than later given Java 7 reached EOL a few months back now. We should also consider associating this change with at least a minor version increase (0.6.0?) given it will cause compatibility issues for clients.

beloglazov commented on build.sbt:
```diff
@@ -10,32 +10,32 @@ name := "couchdb-scala"
 
 version := "0.5.3-SNAPSHOT"
```
As you mentioned, since switching to jdk8 is a significant change, I think we should move to 0.6.0-SNAPSHOT at this stage.

beloglazov commented on build.sbt:
```diff
@@ -10,32 +10,32 @@ name := "couchdb-scala"
+ "org.typelevel"               %% "scalaz-specs2"               % "0.3.0"  % "test",
+ "org.scalacheck"              %% "scalacheck"                  % "1.12.1" % "test",
+ "org.scalaz"                  %% "scalaz-scalacheck-binding"   % "7.1.0"  % "test",
+ "ch.qos.logback"              %  "logback-classic"             % "1.1.2"  % "test"
```
Could you please not change the indentation here and below, it screws up the diffs :)

beloglazov commented on src/main/scala/com/ibm/couchdb/api/Documents.scala:
```diff
@@ -30,40 +30,40 @@ class Documents(client: Client, db: String, typeMapping: TypeMapping) {
   val server = new Server(client)
 
-  private def getClassName[T](obj: T): String = obj.getClass.getCanonicalName
+  private def getClassName[S](obj: S): String = obj.getClass.getCanonicalName
```
Why `S` instead of `T`?

ermyas replied:
with upickle 0.3.x this seems to create a type parameter name conflict, with the following error message "type parameter T defined in method x shadows type T defined in method ...  You may want to rename your type parameter, or possibly remove it.".  The simplest work around was to rename the type parameter

beloglazov replied:
We should keep this type parameter name consistent with other places, what do you think about changing it everywhere to `D` (for Document)? It's already called `D` in https://github.com/beloglazov/couchdb-scala/blob/updateDependencyVersions/src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala#L107

beloglazov commented on src/test/scala/com/ibm/couchdb/spec/Fixtures.scala:
```diff
@@ -18,7 +18,7 @@ package com.ibm.couchdb.spec
 import com.ibm.couchdb.Lenses._
 import com.ibm.couchdb._
-import monocle.macros.Lenser
+import monocle.macros.{GenLens, Lenser}
```
`Lenser` is unused.

beloglazov commented:
Sorry guys, I've been away for a week, it took me a while to get to this PR. I'll merge it in once the comments are addressed.

@ermyas I think switching to jdk8 is reasonable at this point.

bryce-anderson commented:
One note: http4s is on v0.10.0 now but that shouldn't change anything if you do/don't bump the version.

ermyas committed:
updated version to 0.6.0-SNAPSHOT.
Upgraded versions of http4s, upickle and scalaz to their latest releases.
Minor refactoring and formatting changes.

ermyas commented:
thanks @bryce-anderson i've updated to v0.10.0 
@beloglazov comments addressed. Sorry about the earlier formatting changes to the build.sbt, i need to change my auto-formatting config in intellij :-)

beloglazov commented on src/test/scala/com/ibm/couchdb/spec/Fixtures.scala:
```diff
@@ -18,7 +18,7 @@ package com.ibm.couchdb.spec
 import com.ibm.couchdb.Lenses._
 import com.ibm.couchdb._
-import monocle.macros.Lenser
+import monocle.macros.{GenLens}
```
The braces are unnecessary :)

beloglazov commented on src/main/scala/com/ibm/couchdb/api/Documents.scala:
```diff
@@ -30,40 +30,40 @@ class Documents(client: Client, db: String, typeMapping: TypeMapping) {
+  private def getClassName[S](obj: S): String = obj.getClass.getCanonicalName
 
-  def create[T: upickle.Writer](obj: T): Task[Res.DocOk] = {
+  def create[S: upickle.default.Writer](obj: S): Task[Res.DocOk] = {
```
With `default` the names become quite long, what do you think about adding aliases as in https://github.com/beloglazov/couchdb-scala/blob/updateDependencyVersions/src/main/scala/com/ibm/couchdb/core/Client.scala#L34 ? Maybe these aliases can be moved to `UpickleImplicits`, please check if they get automatically picked up from there.

ermyas replied:
upickle already has alias types `R` and `W` for this so I think we can use those instead of creating our own

ermyas committed:
removed unnecessary braces from import statement

ermyas committed:
Minor refactoring for consistency and brevity

- Replaced references to Reader, Writer with their alias types R, W for brevity.
- Changed type parameter name to `D` in place of `T` or `S` where applicable for consistency

beloglazov commented:
@ermyas looks good, thanks!

There are 3 warnings now when you run `sbt test` (introduced in 862e48883658a94361f392eb38954af4f32fabba, I missed them last time):

``` Bash
[warn] src/test/scala/com/ibm/couchdb/api/DocumentsSpec.scala:281:38: Avoid using null
[warn] src/test/scala/com/ibm/couchdb/api/DocumentsSpec.scala:232:10: Public method must have explicit type
[warn] src/test/scala/com/ibm/couchdb/api/DocumentsSpec.scala:241:10: Public method must have explicit type
```

Could you please fix them in another PR? :)
